Update EIP-4762: reworked gas schedule from interop (#8550)
ethereum/EIPs: EIPS/eip-4762.md

#### Access events for account headers

When:

 1. a non-precompile address is the target of a `CALL`, `CALLCODE`, `DELEGATECALL`, `STATICALL`, `SELFDESTRUCT`, `EXTCODESIZE`, or `EXTCODECOPY` opcode, 
 2. a non-precompile address is the target address of a contract creation whose initcode starts execution,
 3. any address is the target of the `BALANCE` opcode

> process this access events:

Author: clarify if `CODESiZE` needs to charge the warm costs

#### Access events for account headers

When:

>  1. a non-precompile address is the target of a `CALL`, `CALLCODE`, `DELEGATECALL`, `STATICALL`, `AUTHCALL`, `SELFDESTRUCT`, `EXTCODESIZE`, or `EXTCODECOPY` opcode, 

Reviewer: Do we want to compress `CALL`, `CALLCODE`, `DELEGATECALL`, `STATICALL`, `AUTHCALL` with a `*CALL` as done in other parts of this EIP?

 2. a non-precompile address is the target address of a contract creation whose initcode starts execution,
 3. any address is the target of the `BALANCE` opcode

>  4. a _deployed_ contract calls `CODECOPY`

Reviewer: If you are in a deployed contract, isn't read access effect for the target already done by the respective `*CALL` making this clarification redundant?

Author: this is meant so that the warm costs are charged

```
 1. a non-precompile address is the target of a `CALL`, `CALLCODE`, `DELEGATECALL`, `STATICALL`, `AUTHCALL`, `SELFDESTRUCT`, `EXTCODESIZE`, or `EXTCODECOPY` opcode, 
 2. a non-precompile address is the target address of a contract creation whose initcode starts execution,
 3. any address is the target of the `BALANCE` opcode
 4. a _deployed_ contract calls `CODECOPY`

If a call is value-bearing (ie. it transfers nonzero wei), whether or not the callee is a precompile, process these two access events:
process this access events:

```
(caller_address, 0, BALANCE_LEAF_KEY)
(callee_address, 0, BALANCE_LEAF_KEY)
(address, 0, BASIC_DATA_LEAF_KEY)
```

When a contract is created, process these access events:
If a `*CALL` or `SELFDESTRUCT` is value-bearing (ie. it transfers nonzero wei), whether or not the callee is a precompile, process these additional two access events:

```
(contract_address, 0, VERSION_LEAF_KEY)
(contract_address, 0, NONCE_LEAF_KEY)

> (contract_address, 0, BALANCE_LEAF_KEY)

Reviewer 2: I'm wondering if converting balance from a `uint256` to a `uint128` might cause breaking changes in some contexts. To be backwards compatible, it also puts languages like Solidity in a position where they will have to treat balances as `uint256` (for e.g. the very commonly used `address.balance` returns a `uint256`.) even though the underlying data structure only contains a uint128. Ethereum's ecosystem also includes a variety of layer-2 solutions, sidechains, and off-chain protocols that may rely on the current uint256 account structure. Changing to uint128 could introduce incompatibilities with these systems.

Not saying we shouldn't do it, and I'm a fan of the simplification & more efficient packed data structure, but just pointing out these potential risks.

Reviewer 3: well if balance can never exceed 16 bytes it doesn't matter since clients stack can still treat it as 32 bytes and only writing to and fro the state is 16 bytes

Author: according to etherscan/coinmarketcap, the total ether supply peaked above 120M, which is $120 \times 10^{18 + 6}$ wei, which corresponds to about 87 bits if my math is correct. To reach a situation where this is a problem, we need to multiply the issuance by $2^{42}$ and this would have to land at the same address.

Reviewer 4: > I'm wondering if converting balance from a `uint256` to a `uint128` might cause breaking changes in some contexts. To be backwards compatible, it also puts languages like Solidity in a position where they will have to treat balances as `uint256` (for e.g. the very commonly used `address.balance` returns a `uint256`.) even though the underlying data structure only contains a uint128. Ethereum's ecosystem also includes a variety of layer-2 solutions, sidechains, and off-chain protocols that may rely on the current uint256 account structure. Changing to uint128 could introduce incompatibilities with these systems.
> 
> 
> 
> Not saying we shouldn't do it, and I'm a fan of the simplification & more efficient packed data structure, but just pointing out these potential risks.

Reviewer 5: I think we should include a mechanism to handle balances larger than 128bit, even if it is not used. The overflow balance could be stored in an 'additional balance leaf' which will be accessed if the 128bit one is maxed out. This makes the gas cost for reading balance dynamic though, which could be another source of contract bugs, though we wouldn't be hitting this in practice on eth mainnet.

> (caller_address, 0, BASIC_DATA_LEAF_KEY)

Reviewer: Considering now that `BASIC_DATA_LEAF_KEY` was read-accessed for `SELFDESTRUCT` by [this rule](https://github.com/ethereum/EIPs/pull/8550/files#diff-cd2aaa222faeb6081a41605386a509b3c4839285cd15f1b0ebe009318707914eR44), maybe we can remove this line?

> (address, 0, CODE_HASH_LEAF_KEY)

Reviewer 3: so we need to remove isEmpty check? 

or we assume that if codehash witness is non null then it isn't empty? and for empty accounts it will come null

Reviewer: Lateral question: the consensus wasn't that we still would have to touch "VERSION" to be forward-compatible? If that's the case, we also need to add here `BASIC_DATA_LEAF_KEY` which can simplify all the questions above (since it comes with all the rest)?

Author: we decided to touch `BASIC_DATA_LEAF_KEY`

> (address, 0, CODE_HASH_LEAF_KEY)

Reviewer: In 4762 we have mixed namings with `CODE_HASH_LEAF_KEY` and `CODEHASH_LEAF_KEY`.
And in 6800 we use `CODE_KECCAK_LEAF_KEY`.

I'd say to change everything to `CODE_HASH_LEAF_KEY` or `CODEHASH_LEAF_KEY`.

Reviewer 6: I think the big deciding factor is: when ethereum completes its shift to sha256 as its primary hash function (or even moves to something more stark-friendly), would this still be the keccak, or would it change (even if for new contracts or EOF only)? If still keccak, then call it CODE_KECCAK, if not, call it CODE_HASH.

Author: using `CODEHASH` for future-proofiness

(contract_address, 0, BASIC_DATA_LEAF_KEY)
(contract_address, 0, CODEHASH_LEAF_KEY)
```

If the `SELFDESTRUCT` opcode is called by some `caller_address` targeting some `target_address` (regardless of whether it’s value-bearing or not), process access events of the form:

```

> (caller_address, 0, BASIC_DATA_LEAF_KEY)

Reviewer: If this rule applies if it's value-bearing or not, and for all cases (i.e: including precompiles), I think we might be able to remove `SELFDESTRUCT` from [this rule](https://github.com/ethereum/EIPs/pull/8550/files#diff-cd2aaa222faeb6081a41605386a509b3c4839285cd15f1b0ebe009318707914eR44)?

Author: if value-bearing + precompile : touch balance, if precompile + not value bearing, don't touch it (just like a call)

> (target_address, 0, BASIC_DATA_LEAF_KEY)

Reviewer: If this rule applies if it's value-bearing or not, and for all cases (i.e: including precompiles), I think we might be able to remove `SELFDESTRUCT` from [this rule](https://github.com/ethereum/EIPs/pull/8550/files#diff-cd2aaa222faeb6081a41605386a509b3c4839285cd15f1b0ebe009318707914eR44)?

And also being mentioned from [this other rule](https://github.com/ethereum/EIPs/pull/8550/files#diff-cd2aaa222faeb6081a41605386a509b3c4839285cd15f1b0ebe009318707914eR55)?

> (target_address, 0, BASIC_DATA_LEAF_KEY)

Reviewer: So, isn't this L74-L78 rule more general than [this one](https://github.com/ethereum/EIPs/pull/8550/files#diff-cd2aaa222faeb6081a41605386a509b3c4839285cd15f1b0ebe009318707914eR55) for `SELFDESTRUCT`?

If that's the case, we could remove `SELFDESTRUCT` mentioned in that linked rule, and only leave this L74-L78?

Reviewer: I'm starting to feel if wouldn't be easier for everyone to simply list rules for each independent instruction. Like `*CALL`, `SELFDESTRUCT`, etc, instead of trying to coalesce overlaps and re-clarify later with further rules some accessed extensions.

Yes, separating rules by instructions might have some redundancy but for implementors and testing it will be entirely clear what to do for each instruction. Now you have to read the whole document to see if in an extra scenario the same instruction you're interested in has another rule.

Reviewer 7: yes, i kind of agree with this. Listing every opCode seperately will be a better idea. 
Right now its much easier to miss something.

Reviewer 8: I agree, it's hard to follow specs like that, I also understand the idea of ​​having a concise EIP

#### Access events for account headers

When:

 1. a non-precompile address is the target of a `*CALL`, `SELFDESTRUCT`, `EXTCODESIZE`, or `EXTCODECOPY` opcode, 
 2. a non-precompile address is the target address of a contract creation whose initcode starts execution,
 3. any address is the target of the `BALANCE` opcode
 4. a _deployed_ contract calls `CODECOPY`

process this access events:

```
(address, 0, BASIC_DATA_LEAF_KEY)
```

If a `*CALL` or `SELFDESTRUCT` is value-bearing (ie. it transfers nonzero wei), whether or not the callee is a precompile, process these additional two access events:

```
(caller_address, 0, BASIC_DATA_LEAF_KEY)
(callee_address, 0, BASIC_DATA_LEAF_KEY)
```

> When calling `EXTCODEHASH`, process the access event:

Reviewer: @gballet, I think we should add here "if the target isn't a precompile".

Author: I don't think so: looking at 2929, there is no check whether the target is a precompile or not.

Reviewer: I believe EIP-2929 rules are because you're supposed to load the whole account in one go.

What this rule is saying in this EIP is that adding a precompile codehash to the witness makes sense, which I think it doesn't? In other parts of this EIP we have rules that exclude precompiles.

Author: ```suggestion
When calling `EXTCODEHASH` on a contract that is not a precompile, process the access event:
```

Author: TODO: add rationale for this

When a contract is created, process these access events:
If a `*CALL` or `SELFDESTRUCT` is value-bearing (ie. it transfers nonzero wei), whether or not the callee is a precompile, process these additional two access events:

```
(contract_address, 0, VERSION_LEAF_KEY)
(contract_address, 0, NONCE_LEAF_KEY)
(contract_address, 0, BALANCE_LEAF_KEY)
(contract_address, 0, CODE_KECCAK_LEAF_KEY)
(contract_address, 0, CODE_SIZE_LEAF_KEY)
(caller_address, 0, BASIC_DATA_LEAF_KEY)
(callee_address, 0, BASIC_DATA_LEAF_KEY)
```

If the `BALANCE` opcode is called targeting some address, process this access event:
When calling `EXTCODEHASH`, process the access event:

```

> (address, 0, BALANCE_LEAF_KEY)

Reviewer: Shouldn't we also touch the `VERSION` thus `BASIC_DATA_LEAF_KEY`?

Author: TODO: add version

> (target_address, 0, BASIC_DATA_LEAF_KEY)

Author: Touching the code hash as well in case the target account is created.

```suggestion
(target_address, 0, BASIC_DATA_LEAF_KEY)
(target_address, 0, CODEHASH_LEAF_KEY)
```

> When a nonzero-balance-sending `CALL` or `SELFDESTRUCT` with a given sender and recipient takes place, process these write events:

Reviewer: Should say `*CALL`?

```
(sender, 0, BASIC_DATA_LEAF_KEY)
(recipient, 0, BASIC_DATA_LEAF_KEY)
```

When a contract creation is initialized, process these write events:

```
(contract_address, 0, BASIC_DATA_LEAF_KEY)
```

When a contract is created, process these write events:

```

> (contract_address, 0, BASIC_DATA_LEAF_KEY)

Reviewer: Is this line needed given that L126 does the same write-access and `initialized < created`?

Author: ```suggestion
```

For a transaction, make these access events:

```
(tx.origin, 0, BASIC_DATA_LEAF_KEY)
(tx.origin, 0, CODEHASH_LEAF_KEY)
(tx.target, 0, BASIC_DATA_LEAF_KEY)
(tx.target, 0, CODEHASH_LEAF_KEY)
```

#### Write events

```
(tx.origin, 0, BASIC_DATA_LEAF_KEY)
```

> if `value` is non-zero:

Reviewer: ```suggestion
If `value` is non-zero:
```

For a transaction, make these access events:

```
(tx.origin, 0, BASIC_DATA_LEAF_KEY)

> (tx.origin, 0, CODEHASH_LEAF_KEY)

Reviewer: If we have a tx (non-value-bearing) targeting a precompile or system contract, we don't need to touch BASIC_DATA. (i.e: stateless client knows the code-size/hash)

We can still argue that this rule should be avoided to avoid a border case. But considering we didn't do any similar simplification for precompiles/system contracts in other places in the EIP (i.e. we avoid adding in the witness for them as much as possible), then maybe not doing it here is actually _adding_ a border case (we have to remember this is an exception), not removing it.

No strong opinions, just surfacing this.
cc @gballet @g11tech @gabrocheleau @tanishqjasoria

> (tx.target, 0, BASIC_DATA_LEAF_KEY)

Author: talk with @jsign: also add the code hash so that it's warmed

Author: note that this collapses the next paragraph

```

> if a transaction is a successful contract creation, make this extra access event before charging the costs for the bytecode storage:

Reviewer 8: could we clarify if it is before this one https://github.com/ethereum/EIPs/blob/29da7466816d023031fe0171e5adc66204fc6d87/EIPS/eip-4762.md?plain=1#L152 ?

Reviewer 8: do we have to access it again even if we did https://github.com/ethereum/EIPs/blob/29da7466816d023031fe0171e5adc66204fc6d87/EIPS/eip-4762.md?plain=1#L72 ?

|`WITNESS_BRANCH_COST`	|1900|
|`WITNESS_CHUNK_COST`	|200|
|`SUBTREE_EDIT_COST`	|3000|
|`CHUNK_EDIT_COST`	    |500|

> |`CHUNK_FILL_COST`	    |6200|

Reviewer: Intended?

> |`CHUNK_FILL_COST`	    |6200|

Reviewer: I can't add this comment in the intended line because there wasn't a change there.
I'm referring to lines L218-L220 and L226.

The intention in both places to say "skip it" is for gas cost charging. But the thing being skipped is also adding it to the underlying access list that are used for the witness creation. If those things also skip adding them to the list, then I think the witness wouldn't be correct?

Author: but the adding of stuff in the access list is done in previous sections, whereas this is specific to the charging of costs.

Note that tree keys can no longer be emptied: only the values `0...2**256-1` can be written to a tree key, and 0 is distinct from `None`. Once a tree key is changed from `None` to not-`None`, it can never go back to `None`.

> Note that values should only be added to the witness if there is sufficient gas to cover their associated event costs.

Reviewer 3: ahh nice

`CREATE*` and `*CALL` reserve 1/64th of the gas before the nested execution. Note that, following the behavior of access lists, the 1/64th of the gas is subtracted:

>  * **BEFORE** charging the witness costs when performing a `CALL`, `CODECALL`, `DELEGATECALL` or`STATICCALL`

Reviewer: Want to use `*CALL` to also contemplate `AUTHCALL`? (Also note a missing space after `or`)

>  * **AFTER** charging th witness costs when performing a `CREATE` or `CREATE2`

Reviewer: ```suggestion
 * **AFTER** charging the witness costs when performing a `CREATE` or `CREATE2`
```

Note that tree keys can no longer be emptied: only the values `0...2**256-1` can be written to a tree key, and 0 is distinct from `None`. Once a tree key is changed from `None` to not-`None`, it can never go back to `None`.

Note that values should only be added to the witness if there is sufficient gas to cover their associated event costs.

`CREATE*` and `*CALL` reserve 1/64th of the gas before the nested execution. In order to match the behavior of this charge with the pre-fork behavior of access lists: 

>  * this 1/64th of the gas is subtracted **AFTER** charging the witness costs when performing a `CALL`, `CODECALL`, `DELEGATECALL` or`STATICCALL`

Reviewer: Nit: in the case of `*CALL` (compared to `CREATE*`) it isn't a strict reservation, but a "cap check" of the max amount of gas you can use. 

If the `*CALL` parameter provided a `gas` parameter that is less 63/64, then _that_ `gas` is reserved.

```suggestion
 * this minimum 1/64th gas reservation is checked **AFTER** charging the witness costs when performing a `CALL`, `CODECALL`, `DELEGATECALL` or`STATICCALL`
```

Anyone reading the spec as it's today will understand the important clarification being made here. So I'm fine if we dismiss this nit.

>  * this 1/64th of the gas is subtracted **AFTER** charging the witness costs when performing a `CALL`, `CODECALL`, `DELEGATECALL` or`STATICCALL`

Reviewer: @gballet, I wonder if we're missing an extra clarification here.

The `*CALL` variants that can be value-bearing, should also respect this rule?

Today in geth, we have this order (e.g: `makeCallVariantGasEIP4762(gasCall)`):
- Charge general `*CALL` costs. (`makeCallVariantGasEIP4762(...)`)
- Do the 1/64 stuff ((`gasCall(...)`)
- Charge BALANCE if `transfersValue` (`gasCall(...)`)

So I'm wondering if this requires a clarification, or was intentional. If was intentional it means that today in kaustinenv6-geth we need a change.

    elements: BitVector[256]
```

## Block-level operations

> * Precompile accounts, system contract accounts and slots of a system contract that are accessed during a system call are considered warm.

Reviewer 3: just to reconfirm this is only for the system call?

also system call is:
 i) system updates not related to a tx
 ii) a precompile executing system call code  (blockhash)

Author: system calls, right now, are 2935 (post osaka), 4788 and 7002. whatever stuff they use that get accessed, should be considered warm.

I am actually close to reconsidering that, it's a special case.

Reviewer: Post-Osaka-2935 has a long history. I'm not sure accesses to that history via the system contract should be considered warm?

Reviewer: From what I understand the intention is here, "system call" refers to any instruction (e.g: BLOCKHASH) or system contract execution called during a tx.

So "system call" has to contemplate both instructions and contracts, thus this is the reason why this concept emerges? (just trying to double-check).

Reviewer 3: > reg your last paragraph, system updates from my perspective doesn't make sense to say are cold or warm, since we can't charge anyone for that?

yes I mean whateever system updates access warms for the txs. With this logic only the last entry in the ring buffer would be warmed and others cold so not the rest 8191 hashes

Author: we decided to not warm anything.

> * Precompile accounts, system contract accounts and slots of a system contract that are accessed during a system call are considered warm.

Reviewer 7: should we add that the code chunks during system calls are not to be warmed? or it might cause additional confusion?

Reviewer 3: system calls totally ignore code chunks imo both for accesses and gas compute and we can clearly mention it to remove confusion

> * When (and only when) calling a system contract _via a system call_, the code chunks and account should not appear in the witness.

Reviewer 3: this implies that stateless clients need to maintain system contracts . Also there is an unresolved issue of stateless clients not knowing if the contract has been actually deployed or not.

we should simply declare that stateless clients should assume activated system contracts deployed.

Author: > this implies that stateless clients need to maintain system contracts 

it does not: because there is no obligation to call the contract's EVM, you are free to write the storage slots directly without having a local copy of the code.

>  Also there is an unresolved issue of stateless clients not knowing if the contract has been actually deployed or not.

I also don't see why this would be true. System contract activation is part of the pectra fork, so it should assume that.

> we should simply declare that stateless clients should assume activated system contracts deployed.

indeed. Or do you mean that it should be in the spec itself?

Reviewer 3: > because there is no obligation to call the contract's EVM, you are free to write the storage slots directly without having a local copy of the code

for system contracts like 7002, even stateless focused clients like ethereumjs just does the EVM call to get the triggered exits  contract to get the exits. But this isn't a deal breaker just something clients need to bundle in verkle.

> System contract activation is part of the pectra fork

deployment is a manual exercise and not really part of client hardfork (that clients deploy this contract on forkblock which imo should have been a better mechanism to deploy the tx if not already deployed). I mean I already had done the mistake of not adding a contract in a local testnet and then debugging why things weren't working. So I would like things foolproof if possible. The problem could become acute as the number of such contracts grow.

> indeed. Or do you mean that it should be in the spec itself?

yes a simple mention of this would clearup the "ruleset" for the stateless client.

> * The coinbase account is warmed before the block, and the whole account is accessed

Reviewer 3: whole account = basic data + code hash?

Author: yes

Author: this is because the coinbase might not exist, so to avoid a lot of issues, we decide to touch the whole account so that it's created if it doesn't exist already.

Reviewer 3: what if there is no tx that puts any value to it, and lets say there is no tips that get added to a non existent coinbase, then it should be 158 cleaned up in the end of block?

> * Withdrawals are cold, but are added to the witness at no cost if they are the target of a transaction.

Reviewer 7: we should add the clarification about storage cells here.
If in a trasnaction, storage value is set and then reset to zero, we should insert zero to the tree and not leave it empty because the user paid the gas cost.
Basically if we charge CHUNK_FILL_COST, fill the chunk value.

Reviewer 3: lateral question/observation to the above:

this is just because withdrawals are processed at the end of the block, and there is no actual cost charged to the coinbase for processing them?
